src/IECore/StringAlgo.cpp

```diff
@@ -95,6 +99,183 @@ bool matchInternal(
+					bool bracketed = *s =='{';
+					const char *variableNameStart = nullptr;
+					const char *variableNameEnd = nullptr;
+					if( bracketed )
+					{
+						s++; // skip initial bracket
+						variableNameStart = s;
+						while( *s && *s != '}' )
+						{
+							s++;
+						}
+						variableNameEnd = s;
+						if( *s )
+						{
+							s++; // skip final bracket
+						}
+					}
+					else
+					{
+						variableNameStart = s;
+						while( isalnum( *s ) )
+						{
+							s++;
+						}
+						variableNameEnd = s;
+					}
+
+					bool recurse = false;
+					const std::string &variable = variables.variable(
+						boost::string_view( variableNameStart, variableNameEnd - variableNameStart ),
+						recurse
+					);
+
+					if( recurse )
+					{
+						substituteInternal( variable.c_str(), variables, result, recursionDepth + 1, substitutions );
+					}
+					else
+					{
+						result += variable;
+					}
+				}
+				else
+				{
+					// variable substitutions disabled
+					result.push_back( *s++ );
+				}
+				break;
+			}
+			case '#' :
+			{
+				if( substitutions & StringAlgo::FrameSubstitutions )
+				{
+					int padding = 0;
+					while( *s == '#' )
+					{
+						padding++;
+						s++;
+					}
+					std::ostringstream padder;
+					padder << std::setw( padding ) << std::setfill( '0' ) << variables.frame();
+					result += padder.str();
+				}
+				else
+				{
+					// frame substitutions disabled
+					result.push_back( *s++ );
+				}
+				break;
+			}
+			case '~' :
+			{
+				if( substitutions & StringAlgo::TildeSubstitutions && result.size() == 0 )
+				{
+					if( const char *v = getenv( "HOME" ) )
+					{
+						result += v;
+					}
+					++s;
+					break;
+				}
+				else
+				{
+					// tilde substitutions disabled
+					result.push_back( *s++ );
+				}
+				break;
+			}
+			default :
+				result.push_back( *s++ );
+				break;
+		}
+	}
+}
+
+struct CompoundDataVariableProvider : public StringAlgo::VariableProvider
+{
+
+	CompoundDataVariableProvider( const CompoundData *variables )
+		:	m_variables( variables )
+	{
+	}
+
+	int frame() const override
+	{
+		const IECore::IntData *d = m_variables->member<IECore::IntData>( "frame" );
+		return d ? d->readable() : 1;
+	}
+
+	const std::string &variable( const boost::string_view &name, bool &recurse ) const override
+	{
+		const IECore::Data *d = m_variables->member<IECore::Data>( name );
+		if( d )
+		{
+			switch( d->typeId() )
+			{
+				case IECore::StringDataTypeId :
+					recurse = true;
+					return static_cast<const IECore::StringData *>( d )->readable();
+				case IECore::FloatDataTypeId :
+					m_formattedString = boost::lexical_cast<std::string>(
+						static_cast<const IECore::FloatData *>( d )->readable()
+					);
+					return m_formattedString;
+				case IECore::IntDataTypeId :
+					m_formattedString = boost::lexical_cast<std::string>(
+						static_cast<const IECore::IntData *>( d )->readable()
+					);
+					return m_formattedString;
+				default :
```

Comment:
In the Context implementation we handle `InternedStringVectorData` (albeit by assuming its a ScenePath) as well as environment variables. Any reason to leave them out here?

Reply:
- InternedStringVectorData : In Gaffer, it's not _too_ much of a stretch to say that InternedStringVectorData in a Context is almost certainly "scene:path" so I was comfortable doing the expedient thing and assuming it should be formatted as such (although note the comment in the code about alternative approaches). In the broader context of Cortex I couldn't make the same justification. It's also fairly trivial to provide your own formatting now, so if you know what you want, you can implement it.
- Environment variables : In Gaffer, we can say "it is useful to end users if we fall through to lookup environment variables", and we can say "we must freeze environment values at startup because otherwise changes will break the cache". We can't say either in the case of a general purpose method in a support library. Again, it's fairly trivial to either do `myVariables.update( os.environ )` in Python or make a small VariableProvider in C++ if you know what behaviour you want.